sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala

```diff
@@ -481,12 +481,12 @@ case class Add(
     // TODO: do not reorder consecutive `Add`s with different `evalMode`
     val reorderResult =
       orderCommutative({ case Add(l, r, _) => Seq(l, r) }).reduce(Add(_, _, evalMode))
-    if (resolved && reorderResult.resolved && reorderResult.dataType == dataType) {
-      reorderResult
+    if (resolved && reorderResult.resolved && reorderResult.dataType != dataType) {
```

Comment:
for safe, can we do this for decimal type only ? e.g
```scala
left match {
  case _: DecimalType if resolved && reorderResult.resolved && reorderResult.dataType != dataType =>
    Cast(reorderResult, dataType)
  case _ => reorderResult
}
```

Reply:
@ulysses-you the current code seems fine. If there is a new data type in the future, we can still avoid the issue.

Reply:
I'm not worry about new data types, just want to avoid unnecessary logic for some unrelated data types like integer, float ..